### PR TITLE
Bug time sync

### PR DESCRIPTION
**Pull request by pushtheworldllc**
# 1.1.0

### New Features

* Add function `.time()` which should be used in time syncing
* Add function `.syncClocksFull()` which should be used for immediate consecutive time syncs
* Synced object can be emitted on `synced` event. Check `valid` property for if the sync was done
* Add detailed description of object returned on `synced` event to README.md

### Breaking Changes

* Changed option named `timeSync` to `sntpTimeSync`
* Removed function called `.sntpNow()` because it was replaced by `.time()`

### Bug Fixes

* Time sync working
* Module could not work with local time

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -162,31 +162,56 @@ Time Syncing
 ------------
 You must be using OpenBCI firmware version 2 in order to do time syncing. After you `.connect()` and send a `.softReset()`, you can call `.usingVersionTwoFirmware()` to get a boolean response as to if you are using `v1` or `v2`.
 
-Now using firmware `v2`, the fun begins! What we set out to do was synchronize not only the board to this modules clock, but also with a global NTP server, so that you could use several different devices and all sync to the same global server. That way you can really do some serious cloud computing!
+Now using firmware `v2`, the fun begins! We synchronize the Board's clock with the module's time. In firmware `v2` we leverage samples with time stamps and different time stamps and unique new features to calculate a time sync strategy. Time syncing has been verified to +/- 4ms and a test report is on the way. We are still working on the synchronize of this module and an NTP server, this is an open call for any NTP experts out there! With a global NTP server you could use several different devices and all sync to the same time server. That way you can really do some serious cloud computing!
```
Comment (teonbrooks) [inline README.md]: `with time stamps and different time stamps`?

File: README.md [hunk 2]
```diff
@@ -162,31 +162,56 @@ Time Syncing
...
+var sampleRate = k.OBCISampleRate250; // Default to 250, ALWAYS verify with a call to `sampleRate` after `ready` event!
+var timeSyncPossible = false;
+
 // Call to connect
 ourBoard.connect(portName).then(() => {
     ourBoard.on('ready',() => {
+        // Get the sample rate after 'ready' event!
+        sampleRate = ourBoard.sampleRate();
+        // Find out if you can even time sync, you must be using v2 and this is only accurate after a `.softReset()` call which is called internally on `.connect()`. We parse the `.softReset()` response for the presence of firmware version 2 properties.
+        timeSyncPossible = ourBoard.usingVersionTwoFirmware();
```
Comment (teonbrooks) [inline README.md]: why not use a convention like the js version of python's `ourBoard.__version__`?

File: README.md [hunk 3]
```diff
@@ -841,6 +856,96 @@ Send the command to tell the board to start the syncing protocol. Must be connec
...
+                    /** Could also call `.syncClocksFull()` here */
+                })
+                .catch(err => {
+                    console.log(`Error starting stream ${err}`);
+                })
+            });
+        ourBoard.on('sample',sample => {
+            /** If we are not synced, then do that! */
+            if (timeSyncActivated === false) {
+                timeSyncActivated = true;
```
Comment (teonbrooks) [inline README.md]: where is this used?

File: README.md [hunk 4]
```diff
@@ -841,6 +856,96 @@ Send the command to tell the board to start the syncing protocol. Must be connec
...
+                        }
+                        return ourBoard.syncClocksFull();
+                    })
+                    .then(syncObj => {
+                        if (syncObj.valid) {
+                            console.log('3rd sync done');
+                        }
+                        return ourBoard.syncClocksFull();
+                    })
+                    .then(syncObj => {
```
Comment (teonbrooks) [inline README.md]: is this just demonstrating how to do four time syncs? the example above had the sync based on the modulus of the sample count, which makes sense.

File: openBCIBoard.js [hunk 5]
```diff
@@ -7,7 +7,7 @@ var EventEmitter = require('events').EventEmitter,
     openBCISample = require('./openBCISample'),
     k = openBCISample.k,
     openBCISimulator = require('./openBCISimulator'),
-    now = require('performance-now'),
+    now = require("performance-now"),
```
Comment (teonbrooks) [inline openBCIBoard.js]: why double quotes?

File: openBCIBoard.js [hunk 6]
```diff
@@ -1534,14 +1535,46 @@ function OpenBCIFactory() {
...
+            this.sync.curSyncObj.timeSyncSent = this.time();
             this.curParsingMode = k.OBCIParsingTimeSyncSent;
             this._writeAndDrain(k.OBCISyncTimeSet);
             resolve();
         });
     };
 
     /**
+     * @description Send the command to tell the board to start the syncing protocol. Must be connected,
+     *      streaming and using version +2 firmware. Uses the `synced` event to ensure multiple syncs
```
Comment (teonbrooks) [inline openBCIBoard.js]: `+2` as opposed to just `2`?

File: openBCIBoard.js [hunk 7]
```diff
@@ -1788,49 +1824,124 @@ function OpenBCIFactory() {
...
+    OpenBCIBoard.prototype._processPacketTimeSyncSet = function(rawPacket, timeOfPacketArrival) {
         return new Promise((resolve, reject) => {
-            var currentTime = this.sntpNow();
+            if (this.sync.curSyncObj === null) reject('no sync in progress');
+            // console.log('hey');
+            this.sync.curSyncObj.timeSyncSetPacket = timeOfPacketArrival;
+
             if (this.options.verbose) console.log('Got time set packet from the board');
+
+            // this.curParsingMode will equal k.OBCIParsingNormal if comma found
```
Comment (teonbrooks) [inline openBCIBoard.js]: I thought a comma was expected here. or is the `>` expected?

File: openBCIBoard.js [hunk 8]
```diff
@@ -1788,49 +1824,124 @@ function OpenBCIFactory() {
...
+                    // Calculate the time between sending the `<` to getting the set packet, call this the round trip length
+                    this.sync.curSyncObj.timeRoundTrip = this.sync.curSyncObj.timeSyncSetPacket - this.sync.curSyncObj.timeSyncSent;
+                    if (this.options.verbose) console.log(`Round trip time: ${this.sync.curSyncObj.timeRoundTrip} ms`);
+
+
+                    // If the sync sent conf and set packet arrive in different serial flushes
+                    //  ------------------------------------------
+                    // |       |        timeTransmission          |  < GOOD :)
+                    //  ------------------------------------------
+                    // ^       ^                                  ^
```
Comment (teonbrooks) [inline openBCIBoard.js]: the `sent` is no longer vertically aligned

File: openBCIBoard.js [hunk 9]
```diff
@@ -1788,49 +1824,124 @@ function OpenBCIFactory() {
...
+                    //  ------------------------------------------
+                    // |                                   |      |  < BAD :(
+                    //  ------------------------------------------
+                    // ^                                   ^      ^
+                    //  s                                   s      s
+                    //   e                                   e      e
+                    //    n                                   n      t packet
+                    //     t                                   t confirmation
+                    if ((this.sync.curSyncObj.timeSyncSetPacket - this.sync.curSyncObj.timeSyncSentConfirmation) < k.OBCITimeSyncThresholdTransFailureMS) {
+                        // Estimate that 75% of the time between sent and set packet was spent on the packet making it's way from board to this point
```
Comment (teonbrooks) [inline openBCIBoard.js]: it's --> its

File: openBCIBoard.js [hunk 10]
```diff
@@ -1788,49 +1824,124 @@ function OpenBCIFactory() {
...
+
+
+                    // If the sync sent conf and set packet arrive in different serial flushes
+                    //  ------------------------------------------
+                    // |       |        timeTransmission          |  < GOOD :)
+                    //  ------------------------------------------
+                    // ^       ^                                  ^
+                    //  s      s                                   s
+                    //   e      e                                   e
+                    //    n      n                                   t packet
```
Comment (teonbrooks) [inline openBCIBoard.js]: `t packet` --> `n`. roll this line down a row

Comment (teonbrooks) [inline openBCIBoard.js]: oops, nvm

File: openBCIBoard.js [hunk 11]
```diff
@@ -1949,62 +2061,62 @@ function OpenBCIFactory() {
...
      * @returns {Promise} - A promise if the module was able to sync with ntp server.
      * @author AJ Keller (@pushtheworldllc)
      */
-    OpenBCIBoard.prototype.sntpStart = function() {
+    OpenBCIBoard.prototype.sntpStart = function(options) {
         return new Promise((resolve, reject) => {
-            Sntp.start((err) => {
+            this.options.sntpTimeSync = true;
+            Sntp.start({
+                host: this.options.sntpTimeSyncHost,    // Defaults to pool.ntp.org
```
Comment (teonbrooks) [inline openBCIBoard.js]: does host belong `localhost` if sntp is False?

**Comment by teonbrooks** [comment]
@pushtheworldllc your code is beautiful, man. just left some comments throughout

File: README.md (hunk 2 above)
Comment (pushtheworldllc) [inline README.md]: i wonder what this is for js

File: README.md (hunk 4 above)
Comment (pushtheworldllc) [inline README.md]: yes, added sentence to explain that

File: openBCIBoard.js (hunk 5 above)
Comment (pushtheworldllc) [inline openBCIBoard.js]: double quotes are so hot right now... na good point, i changed back

File: openBCIBoard.js (hunk 7 above)
Comment (pushtheworldllc) [inline openBCIBoard.js]: `<` is sent to dongle and dongle sends `,` back to driver when it actually sends out that `<` to the Board

File: openBCIBoard.js (hunk 11 above)
Comment (pushtheworldllc) [inline openBCIBoard.js]: no you just use `Date.now()` completely avoiding the `Sntp` package

**Pull request merged**